sw/device/boot_rom/bootstrap.c

```diff
@@ -51,66 +52,84 @@ static int erase_flash(void) {
 }
 
 /* Processes frames received via spid interface and writes them to flash. */
 static int bootstrap_flash(void) {
-  static frame_t f;
-  static uint8_t ack[32] = {0};
-  uint32_t expected_frame_no = 0;
-  for (;;) {
-    if (spid_bytes_available() >= sizeof(f)) {
-      spid_read_nb(&f, sizeof(f));
-      uart_send_str("Processing frame no: ");
-      uart_send_uint(f.hdr.frame_num, 32);
-      uart_send_str(" exp no: ");
-      uart_send_uint(expected_frame_no, 32);
-      uart_send_str("\r\n");
-
-      if (FRAME_NO(f.hdr.frame_num) == expected_frame_no) {
-        if (check_frame_hash(&f)) {
-          uart_send_str("Error: detected hash mismatch on frame: ");
-          uart_send_uint(f.hdr.frame_num, 32);
-          uart_send_str("\r\n");
-          spid_send(ack, sizeof(ack));
-          continue;
-        }
-
-        hw_SHA256_hash(&f, sizeof(f), ack);
-        spid_send(ack, sizeof(ack));
-
-        if (expected_frame_no == 0) {
-          flash_default_region_access(/*rd_en=*/1, /*prog_en=*/1,
-                                      /*erase_en=*/1);
-          int rv = erase_flash();
-          if (rv) {
-            return rv;
-          }
-        }
-        if (flash_write(f.hdr.flash_offset, f.data, ARRAYSIZE(f.data))) {
-          return E_BS_WRITE;
-        }
-
-        ++expected_frame_no;
-        if (f.hdr.frame_num & FRAME_EOF_MARKER) {
-          break;
-        }
-      } else {
-        // Send previous ack if unable to verify current frame.
-        spid_send(ack, sizeof(ack));
+  uint8_t ack[32] = {0};
+  uint32_t expected_frame_num = 0;
+  while (true) {
+    size_t bytes_available;
+    (void)dif_spi_device_rx_pending(&spi, &bytes_available);
+    if (bytes_available < sizeof(frame_t)) {
+      continue;
+    }
+
+    size_t bytes_recved;
+    frame_t frame;
+    (void)dif_spi_device_recv(&spi, &frame, sizeof(frame), &bytes_recved);
+    LOG_INFO("Processing frame #%d, expecting #%d\r\n", frame.hdr.frame_num, expected_frame_num);
+
+    if (FRAME_NO(frame.hdr.frame_num) != expected_frame_num) {
+      (void)dif_spi_device_send(&spi, ack, sizeof(ack), NULL);
+      continue;
+    }
+
+    if (!check_frame_hash(&frame)) {
+      (void)dif_spi_device_send(&spi, ack, sizeof(ack), NULL);
+      continue;
+    }
+
+    hw_SHA256_hash(&frame, sizeof(frame_t), ack);
+    (void)dif_spi_device_send(&spi, ack, sizeof(ack), NULL);
+
+    if (expected_frame_num == 0) {
+      flash_default_region_access(/*rd_en=*/1, /*prog_en=*/1,
+                                    /*erase_en=*/1);
+      int error = erase_flash();
+      if (error != 0) {
+        return error;
       }
     }
+
+    if (flash_write(frame.hdr.flash_offset, frame.data, ARRAYSIZE(frame.data)) != 0) {
+      return E_BS_WRITE;
+    }
+
+    ++expected_frame_num;
+    if ((frame.hdr.frame_num & FRAME_EOF_MARKER) != 0) {
+      break;
+    }
   }
-  uart_send_str("bootstrap: DONE!\r\n");
+  LOG_INFO("Bootstrap done!\r\n");
```

Comment:
If you change this you have to update `getting_started_fpga.md` and `quickstart.md` to reflect the change.